### ISSUE TITLE
[svast] Expression nodes start and end can be ambiguous

### ISSUE THREAD

**Issue by halfnelson** [pngwn/MDsveX#174]
Currently when I have an instance of a `SvelteExpression` node, the `start` and `end` positions can be ambiguous.

for plain mustache tag
```svelte
<h1>{title}</h1>
```
I get
```json
      "type": "svelteElement",
      "tagName": "h1",
      "properties": [],
      "selfClosing": false,
      "children": [
        {
          "type": "svelteExpression",
          "value": "title",
          "position": {
            "start": {
              "line": 1,
              "column": 5,
              "offset": 4
            },
            "end": {
              "line": 1,
              "column": 12,
              "offset": 11
            }
          }
        }
      ]
```
The characters of the source represented by position.start ->position.end are `{title}`

But if the expression is a child of a void element
```svelte
<h1>{@debug title}</h1>
```
I get
```json
          "type": "svelteVoidBlock",
          "name": "debug",
          "expression": {
            "type": "svelteExpression",
            "value": "title",
            "position": {
              "start": {
                "line": 1,
                "column": 13,
                "offset": 12
              },
              "end": {
                "line": 1,
                "column": 18,
                "offset": 17
              }
            }
          },
```
With the characters of the source represented by position.start to position.end `title` without the curly.

I can detect this situation from a single `svelteExpression` by checking the length of the `value` with the difference in `offsets`, if they match then we are not a bare mustache tag.

Maybe you could treat `{blah}` as a `svelteVoidBlock` without a name, or with empty string name

**Comment by pngwn** [comment]
Do you think the expression positions should exclude the curlies and just refer to the expression within?

**Comment by halfnelson** [comment]
Not if they aren't then going to be wrapped in a `SvelteVoidBlock`, so that there is still a nice continuous unbroken path from .end to .start through the document.

**Comment by pngwn** [comment]
Ah I think I understand what you mean now. The only problem with that is the bare expression node is used in attributes values and other places as well so the semantics of the `SvelteVoidBlock` don't make as much sense there but on the same token it would be nice to have some consistency with expressions inside blocks.

**Comment by halfnelson** [comment]
true, Maybe a `svelteMoustache` Node which is like a literal except has an expression

**Comment by halfnelson** [comment]
happy to workaround this for now

**Comment by pngwn** [comment]
Re-opened this, it needs to be dealt with. The ambiguity doesn't make sense in an AST.

**Comment by pngwn** [comment]
`{ ... }` will become a `dynamicText` node which contains an expression:

```ts
interface SvelteDynamicContent {
  expression: SvelteExpression;
  position?: Position
}
```

This will only apply to attribute values, raw expression nodes and directive value (the semantic make sense there too). I have no intention of generating a CST (for the foreseeable future at least) so this gives a little bit of flexibility as the AST node position are contiguous showing a relatively full picture of the source, without any ambiguity over positions. 

Expressions in `BranchingBlock` nodes will remain as they are.

**Comment by pngwn** [comment]
This should be fixed in `svelte-parse@0.1.0`. 

I've added a new [`SvelteDynamicContent`](https://github.com/pngwn/MDsveX/blob/master/packages/svast/index.d.ts#L62-L65) node which marks the start and end of the curlies, this node _contains_ the expression node which now has consistent start and end positions.

Let me know if anything is amiss, i had to refactor quite a bit to get this working as well as updating a lot of tests and I could have missed something.
